torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -374,6 +374,361 @@ def collect_bw_donated_buffer_idxs(
+
+                # Save the number of symints and saved tensors
+                new_fw_out_nodes = new_fw_hop_gm.graph.find_nodes(op="output")[0].args[
+                    0
+                ]
+                extra_outputs = new_fw_out_nodes[num_fw_outputs:]
+                symint_outputs = [n for n in extra_outputs if is_sym_node(n)]
+
+                new_hop_graphs[identifier].new_num_sym_nodes = len(symint_outputs)
+                new_hop_graphs[identifier].new_num_saved_nodes = len(
+                    extra_outputs
+                ) - len(symint_outputs)
+
+                new_hop_graphs[identifier].partitioning_done = True
+
+    if not new_hop_graphs:
+        return joint_gm
+
+    # Step 3) Restitch the new fw and bw graphs back into the main graph.
+    #
+    # This is a very mechanical process. There are a quite of pieces that we
+    # need to connect together to make it work. Lets try to understand the
+    # problem statement first.
+    #
+    # For the forward graph, the signature of the old_fw_hop_gm is
+    #   inputs - (*primals)
+    #   outputs - (*fw_outs)
+    # Now the signature of the new_fw_hop_gm is
+    #   inputs - (*primals)     -- This is same
+    #   outputs - (*fw_outs, *saved_tensors)    - This is different
+    # At a high level, this is an easy transformation, in the new graph we just
+    # have to replace the old_fw_hop_gm with the new_fw_hop_gm. Everything else
+    # falls into place, because the input signature (i.e. args) is same. And
+    # even though output signature is different, fw_outs are still at the same
+    # indexes as before. So the forward of the `joint_gm` works nicely.
+    #
+    # Now, lets look at the backward hop graph. Old signature
+    #   inputs - (*primals, *tangents)
+    #   outputs - (*grad_outs, *fw_outs)
+    # New signature
+    #   inputs - (*saved_tensors, *tangents) -- Different
+    #   outputs - (*grad_outs)  -- Different
+    # Here both input and output signature change. The output signature handling
+    # is quite easy because the grads_out are sitting at the right place, so we
+    # dont have to do anything.
+    #
+    # For the input signature, we have to collect the saved tensors from the
+    # corresponding forward graph output. We collect all saved_tensors when we
+    # see the forward graph, and save it into a map and then later use it during
+    # the backward.
+
+    env = {}  # type: ignore[var-annotated]
+    new_graph = torch.fx.Graph()
+
+    already_added_new_hop_mods = set()
+
+    # The stack of fw_nodes for invoke_subgraph HOP. There is an implicit
+    # assumption about the graph structure, i.e., if we have hop1, hop2, hop3,
+    # ... in the forward part of the joint graph, we will have .., hop3, hop2,
+    # hop1 order for the backward. This structure allows us to just use a stack
+    # to collect all the information that we need to pass from the forward hop
+    # node to the corresponding backward node.
```

Comment:
I guess one downside is that this is technically not an invariant today, right? (inductor in theory is free to reorder `invoke_subgraph` calls around each other, although it probably won't right now)

Reply:
Yeah, one can write a joint graph pass that can move hops around. But typically, those reordering happen post partitioning in `compile_fx_inner` functions.